fed/cleanup.py

```diff
@@ -23,97 +23,86 @@
 
 logger = logging.getLogger(__name__)
 
-_sending_obj_refs_q = None
 
-_check_send_thread = None
-
-_EXIT_ON_FAILURE_SENDING = False
-
-
-def set_exit_on_failure_sending(exit_when_failure_sending: bool):
-    global _EXIT_ON_FAILURE_SENDING
-    _EXIT_ON_FAILURE_SENDING = exit_when_failure_sending
-
-
-def get_exit_when_failure_sending():
-    global _EXIT_ON_FAILURE_SENDING
-    return _EXIT_ON_FAILURE_SENDING
-
-
-def _check_sending_objs():
-    def _signal_exit():
-        os.kill(os.getpid(), signal.SIGTERM)
-
-    global _sending_obj_refs_q
-    if not _sending_obj_refs_q:
-        _sending_obj_refs_q = deque()
-
-    while True:
-        try:
-            obj_ref = _sending_obj_refs_q.popleft()
-        except IndexError:
-            time.sleep(0.5)
-            continue
-        if isinstance(obj_ref, bool):
-            break
-        try:
-            res = ray.get(obj_ref)
-        except Exception as e:
-            logger.warn(f'Failed to send {obj_ref} with error: {e}')
-            res = False
-        if not res and get_exit_when_failure_sending():
-            logger.warn('Signal self to exit.')
-            _signal_exit()
-            break
-
-    logger.info('Check sending thread was exited.')
-    global _check_send_thread
-    _check_send_thread = None
-    logger.info('Clearing sending queue.')
-    _sending_obj_refs_q = None
-
-
-def _main_thread_monitor():
-    main_thread = threading.main_thread()
-    main_thread.join()
-    notify_to_exit()
-
-
-_monitor_thread = None
-
-
-def _start_check_sending():
-    global _sending_obj_refs_q
-    if not _sending_obj_refs_q:
-        _sending_obj_refs_q = deque()
-
-    global _check_send_thread
-    if not _check_send_thread:
-        _check_send_thread = threading.Thread(target=_check_sending_objs)
-        _check_send_thread.start()
+class CleanupManager:
+    """
+    This class is used to manage the related works when the fed driver exiting.
+    It monitors whether the main thread is broken and it needs wait until all sending
+    objects get repsonsed.
+
+    The main logic path is:
+        A. If `fed.shutdown()` is invoked in the main thread and every thing works well,
+        the `graceful_stop()` will be invoked as well and the checking thread will be
+        notifiled to exit gracefully.
+
+        B. If the main thread are broken before sending the notification flag to the
+        sending thread, the monitor thread will detech that and it joins until the main
+        thread exited, then notifys the checking thread.
+    """
+
+    def __init__(self) -> None:
+        # `deque()` is thread safe on `popleft` and `append` operations.
+        # See https://docs.python.org/3/library/collections.html#deque-objects
+        self._sending_obj_refs_q = deque()
+        self._check_send_thread = None
+        self._monitor_thread = None
+        self._lock_on_sending_q = threading.Lock()
```

Comment:
`_lock_on_sending_q` and `_lock_on_send_thread` are not used.

Reply:
good catch!